File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def parse_file_name(file):
    i = 0
    head = ""
    number = ""
    while i < len(file):
        v: str = file[i]
        if v.isalpha():
            head += v.lower()
        elif v in " .-":
            head += v
        else:
            break
        i += 1
    while i < len(file):
        v = file[i]
        if v.isdigit():
            number += v
        else:
            break
        i += 1
    if i == len(file) - 1:
        tail = ""
    else:
        tail = file[i:]
    return file, head, int(number), tail


def solution(files):
    files = [parse_file_name(file) for file in files]
    files.sort(key=lambda x: (x[1], x[2]))
    return [file[0] for file in files]

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.Arrays;

class Solution {
    public String[] solution(String[] files) {
        Arrays.sort(files, (a, b) -> {
            String[] aSplit = splitFileName(a);
            String[] bSplit = splitFileName(b);

            String aHead = aSplit[0].toLowerCase();
            String bHead = bSplit[0].toLowerCase();

            if (aHead.equals(bHead)) {
                int aNumber = Integer.parseInt(aSplit[1]);
                int bNumber = Integer.parseInt(bSplit[1]);

                return aNumber - bNumber;
            }

            return aHead.compareTo(bHead);
        });

        return files;
    }

    public String[] splitFileName(String file) {
        int numberStart = -1;
        int numberEnd = -1;
        for (int i = 0; i < file.length(); i++) {
            if (numberStart == -1 && Character.isDigit(file.charAt(i))) {
                numberStart = i;
            }

            if (numberStart != -1 && !Character.isDigit(file.charAt(i))) {
                numberEnd = i - 1;
                break;
            }
        }

        if (numberEnd == -1) {
            numberEnd = file.length() - 1;
        }

        return new String[]{
                file.substring(0, numberStart),
                file.substring(numberStart, numberEnd + 1),
                file.substring(numberEnd + 1),
        };
    }
}